File: q0005/q0005.java
package q0005;

class Solution {
    public String longestPalindrome(String s) {
        if (s.length() == 0)
            return "";
        int l = 0;
        int r = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expandFromCenter(s, i, i);
            int len2 = expandFromCenter(s, i, i + 1);
            int ans = Math.max(len1, len2);
            if (ans > r - l) {
                l = i - (ans - 1) / 2;
                r = i + ans / 2;
            }
        }
        return s.substring(l, r + 1);
    }

    private int expandFromCenter(String s, int l, int r) {
        if (s.length() == 0 || l > r)
            return 0;
        while (l >= 0 &&
                r < s.length() &&
                s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }
        return r - l - 1;
    }
}

File: q0005/q0005.py
class Solution(object):
    def longestPalindrome(self, s):
        if not s:
            return ''

        l = 0
        r = 0

        for i in range(len(s)):
            len1 = self.expandFromCenter(
                s, i, i
            )  ##deals with the case of "racecar", centered around one unique char
            len2 = self.expandFromCenter(s, i, i + 1)  ##deals with the case of "abba"
            lenF = max(len1, len2)
            ## this means we have found new max len palindrome
            if lenF > r - l:
                l = i - ((lenF - 1) // 2)
                r = i + (lenF // 2)
        return s[l : r + 1]

    def expandFromCenter(self, s, l, r):
        if not s or l > r:
            return 0
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        return r - l - 1
